source/framework/core/inc/TRestDataSet.h

```diff
@@ -0,0 +1,103 @@
+ * (at your option) any later version.                                   *
+ *                                                                       *
+ * REST is distributed in the hope that it will be useful,               *
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of        *
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the          *
+ * GNU General Public License for more details.                          *
+ *                                                                       *
+ * You should have a copy of the GNU General Public License along with   *
+ * REST in $REST_PATH/LICENSE.                                           *
+ * If not, see https://www.gnu.org/licenses/.                            *
+ * For the list of contributors see $REST_PATH/CREDITS.                  *
+ *************************************************************************/
+
+#ifndef REST_TRestDataSet
+#define REST_TRestDataSet
+
+#include "ROOT/RDataFrame.hxx"
+#include "TRestMetadata.h"
+#include "TTimeStamp.h"
+
+/// It allows to group a number of runs that satisfy given metadata conditions
+class TRestDataSet : public TRestMetadata {
+   private:
+    /// All the selected runs will have a starting date after fStartTime
+    std::string fStartTime = "2000/01/01";  //<
+
+    /// All the selected runs will have an ending date before fEndTime
+    std::string fEndTime = "3000/12/31";  //<
+
+    /// A glob file pattern that must be satisfied by all files
+    std::string fFilePattern = "";  //<
+
+    /// It contains a list of the observables that will be added to the final tree or exported file
+    std::vector<std::string> fObservablesList;  //<
+
+    /// It contains a list of the process where all observables should be added
+    std::vector<std::string> fProcessObservablesList;  //<
+
+    /// A list of metadata members where filters will be applied
+    std::vector<std::string> fFilterMetadata;  //<
+
+    /// If not empty it will check if the metadata member contains the value
+    std::vector<std::string> fFilterContains;  //<
+
+    /// If the corresponding element is not empty it will check if the metadata member is greater
+    std::vector<Double_t> fFilterGreaterThan;  //<
+
+    /// If the corresponding element is not empty it will check if the metadata member is lower
+    std::vector<Double_t> fFilterLowerThan;  //<
+
+    /// The total integrated run time of selected files
+    Double_t fTotalDuration = 0;  //<
+
+    /// The resulting RDataFrame object after initialization
+    ROOT::RDataFrame fDataSet = 0;  //!
+
+    /// A pointer to the generated tree
+    TTree* fTree = 0;  //!
+
+    /// A list populated by the FileSelection method using the conditions of the dataset
+    std::vector<std::string> fFileSelection;  //!
+
+    void InitFromConfigFile() override;
+
+   protected:
+    virtual std::vector<std::string> FileSelection();
+
+   public:
+    ROOT::RDataFrame GetDataFrame() {
```

Comment:
I am not sure but this may be problematic, we are returning a new `ROOT::RDataFrame` object, instead of a reference to the dataset (which I assume is the intention).

Maybe what we want is

```suggestion
    ROOT::RDataFrame& GetDataFrame() {
```

instead?

It may not be a problem at the end because I think `RDataFrame` does not own the data, but I am not sure.

Reply:
I think we can just accept the new version with the reference. I don't think it makes a copy, clearly.

For me it is ok, but why I cannot press the button commit suggestion?

Reply:
I think you should avoid to return a reference you can either return a pointer or a copy, note that to return a copy is generally optimized.
https://stackoverflow.com/questions/2939165/c-get-method-returning-by-value-or-by-reference